cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java

```diff
@@ -0,0 +1,180 @@
+            Table table = LoggingStoreTableUtil.getMetadataTable(datasetFramework, context);
+            final Row cols = table.get(getRowKey(logPathIdentifier));
+
+            if (cols.isEmpty()) {
+              //noinspection unchecked
+              return new ArrayList<>();
+            }
+
+            final List<LogLocation> files = new ArrayList<>();
+
+            for (Map.Entry<byte[], byte[]> entry : cols.getColumns().entrySet()) {
+              // old rowkey format length is 8 bytes (just the event timestamp is the column key)
+              // new row-key format length is 17 bytes (version_byte, event timestamp (8b) and current timestamp (8b)
+              if (entry.getKey().length == 8) {
+                long eventTimestamp = Bytes.toLong(entry.getKey());
+                if (!(eventTimestamp > endTimestampMs)) {
+                  // old format
+                  files.add(new LogLocation(LogLocation.VERSION_0,
+                                            eventTimestamp,
+                                            // use 0 as current time as this information is not available
+                                            0,
+                                            rootLocationFactory.create(new URI(Bytes.toString(entry.getValue()))),
+                                            logPathIdentifier.getNamespaceId(), impersonator));
+                }
+              } else if (entry.getKey().length == 17) {
+                long eventTimestamp = Bytes.toLong(entry.getKey(), 1, Bytes.SIZEOF_LONG);
+                if (!(eventTimestamp > endTimestampMs)) {
+                  // new format
+                  files.add(new LogLocation(LogLocation.VERSION_1,
+                                            // skip the first (version) byte
+                                            eventTimestamp,
+                                            Bytes.toLong(entry.getKey(), 9, Bytes.SIZEOF_LONG),
+                                            rootLocationFactory.create(new URI(Bytes.toString(entry.getValue()))),
+                                            logPathIdentifier.getNamespaceId(), impersonator));
+                }
+              }
+            }
+            return files;
+          }
+        });
+
+      // performing extra filtering (based on start timestamp) outside the transaction
+      return getFilesInRange(endTimestampFilteredList, startTimestampMs);
+    } catch (TransactionFailureException e) {
+      throw Transactions.propagate(e, Exception.class);
+    }
+  }
+
+  private byte[] getRowKey(LogPathIdentifier logPathIdentifier) {
+    return Bytes.add(ROW_KEY_PREFIX, logPathIdentifier.getRowKey().getBytes());
+  }
+
+  @VisibleForTesting
+  List<LogLocation> getFilesInRange(List<LogLocation> files, long startTimeInMs) {
+    // return if its empty
+    if (files.isEmpty()) {
+      return files;
+    }
+    // sort the list
+    Collections.sort(files, LOG_LOCATION_COMPARATOR);
```

Comment:
Isn't it sorted already when reading from the backing store? The columns are of fixed length and is having (exclude the constant version byte) `eventTime` as the first 8 bytes, followed by the `createTime`